### ISSUE TITLE
`nix-build` is not freeing eval memory during long builds

### ISSUE THREAD

**Issue by andir** [NixOS/nix#5200]
**Describe the bug**

On a fairly simple NixOS system `nix-build` currently allocates about 1.8 - 2GiB of RAM during system config eval. The memory usage is not great but not the topic of this issue. The problem I am seeing is that on low-end devices those 2GiB of RAM that are being allocated to eval the system configuration are a significant portion of the systems resources.

**Steps To Reproduce**

1. Run `nix-build foo.nix` where `foo.nix` defines a NixOS system.
2. Observe the memory growth of the `nix-build` process while the eval is running.
3. As soon as it starts building paths (even just some symlinks, configuration files, ...) it should have free'ed the eval memory but doesn't.
4. Depending on the nix expression and system config you might run out of memory if your system has <4GiB of total RAM available.

**Expected behavior**

The eval memory should not be allocated anymore once the actual build starts.

**`nix-env --version` output**

2.3.15 & 2.4.pre20210802_47e96bb

**Additional context**

`nix-build` doesn't require all of the eval data to stay in memory while running the build. From looking at the code it seems like it needs a list of drv paths and output mappings. None of the inputs to `buildPaths` in the `nix-build.cc` file require the actual eval state to be still around.

Since our current eval situation isn't ideal (with regards to GC, memory leaks, ...) there are generally three solutions I can see:

a) optimize the eval code such that more of the memory is actually GC tracked and release the eval state, the DrvInfos, Expressions lists and ensure that `GC_gcollect` is called once before starting the build.
b) Move all the evalulation into a child process that is forked off from `nix-build` and returns a list of target (output paths, drvs, output mappings, ...). This way the memory consumption will be tied to the lifetime of the children and we don't really have to care about it at all. By the time the build start the eval memory will be free again.
c) Do nothing and tell users to deploy (more) swap such that the memory can be swapped to disk instead of staying in RAM during the entire build. Less ideal but something that can be done today.

I've attempted a) and b) to varying degrees. While refactoring the way our eval code creates lists of expressions (and other currently untracked allocations) has resulted in a few hundred megabytes less memory consumption it introduced random memory failures. Apparently the GC is a bit too eager in some cases or I've missed a spot where we manually free memory.

With b) I've probably not spent enough time tracking down all the global state that has to be restored/closed before/after the child is spawned. It worked well for some simple `nix-shell` invocations but failed during build of an entire NixOS system.

Regarding c) I am not actually sure the memory is inactive enough to be swapped. I haven't tested that theory but I'd assume that this works in most of the cases.

**Comment by edolstra** [comment]
Option b) is more-or-less what we used to have, back when `nix-build` was a Perl script that called `nix-instantiate` and `nix-store -r`. But obviously that's pretty ugly.

**Comment by Pamplemousse** [comment]
> Something that @Pamplemousse did when working on a Nix fuzzer was to allocate all the eval-related values in an arena and free the arena at the end of the evaluation.

Yes; For reference, here is what it looks like: https://github.com/Pamplemousse/nix/commit/f041842ec3294f3bfc5432e26b29d19456c0da79 .
One thing to note though, is that some mutating state was still falling outside of the arena.

> I have no idea whether this can be done while using boehm (in his use-case, the gc was disabled altogether), and that probably requires a bit of work to get right.

I haven't spent too much time tweaking Boehm, but maybe using a combinaison of `GC_MALLOC`, `GC_FREE`, `GC_gcollect` (see https://www.hboehm.info/gc/gcinterface.html) to manage the arena could provide the intended behavior.

> I've probably not spent enough time tracking down all the global state that has to be restored/closed before/after the child is spawned.

@andir, this is something that could greatly benefit fuzzing:
In its current state, the parsing / evaluation part of `nix` **cannot** be fuzzed in memory because of some mutable global state (that I had a hard time to locate...).

**Comment by brandonros** [comment]
There's basically no viable way to build a postgrest armv7l binary at the moment with Nix or Docker, is there?

I guess running postgrest on a Raspberry Pi isn't a very normal usecase?

**Comment by andir** [comment]
> There's basically no viable way to build a postgrest armv7l binary at the moment with Nix or Docker, is there?
> 
> I guess running postgrest on a Raspberry Pi isn't a very normal usecase?

how is this related?

**Comment by roberth** [comment]
I think "a. Make better use of the GC" is still actionable.

<sub>(I've added headers to the description for convenience)</sub>

**Comment by nh2** [comment]
I want to point out that my 4 GiB RAM computer OOM-crashes sometimes during on `nixos-rebuild switch` due to this issue.

Memory usage of `nix-build <nixpkgs/nixos> --no-out-link -A system` balloons up to `1214M` during eval, and does not free itself after eval. Then, depending on what gets built after, `nix-build` gets OOM-killed.

When I `nix-store -r` the system `.drv` that `nixos-rebuild` did output, that only takes 34 MB, and so the machine has another 1.2 GB more RAM available and does not OOM.

So this bug kind-of makes NixOS very unstable on "normal" machines with 4 GB RAM, which is a bit of a shame.

**Comment by sudocurse** [comment]
FWIW 12 GB of swap did not help my build

**Comment by xokdvium** [comment]
Tbh the best thing we could do here is to re-exec ourselves. The OS *is* the best garbage collector out there.

**Comment by xokdvium** [comment]
Specifically, what I'm thinking of is having some special/reserved `argv[0]` values that would get forwarded to the required functions. By re-execing we start with a clean slate and the whole heap would get garbage collected in an efficient manner. Keeping the store paths for derivations in environment variables should be sufficient for the GC (the nix store GC) to keep those alive as roots.

@andir does that make sense to you? This would also significantly improve build performance in the current architecture when not going through the daemon, where the builder processes are `fork`-ed directly from the nix cli process or the daemon process (which might have a gigantic heap and lead to incredibly slow fork() - on the order of 100s of milliseconds). This of course could be further improved by doing more `vfork`-style things or a zygote process, but both of those are complicated and are only tangentially related to this issue.

**Comment by nh2** [comment]
> Tbh the best thing we could do here is to re-exec ourselves. The OS _is_ the best garbage collector out there.

@xokdvium I don't understand why this would be needed. Nix is written in C++, a language very good at controlling memory. Why would it have to `exec()` in order to implement what looks like a normal `free()`?

**Comment by xokdvium** [comment]
> Why would it have to exec() in order to implement what looks like a normal free()?

Boehm GC has no way to nuke the whole heap, and doing mark/sweep stop-the-world GC is expensive + we'd have to be sure that we don't accidentally persist any pointers to GC memory. That's not to say that it can't be done, but it going to be more expensive than throwing away the whole address space which we really don't need anymore most of the time.

That is how nix-eval-jobs works too and we should probably do more of this type of thing.

**Comment by xokdvium** [comment]
> Nix is written in C++, a language very good at controlling memory.

To be clear, most of the memory that's allocated by the evaluation:

- Either is allocated with Boehm and is not actually managed by nix at all. Reclamation/deallocation etc is done by Boehm when it sees fit after doing the marking.

- Is not reclaimable at all – some data structures are append-only.

**Comment by nh2** [comment]
@xokdvium I am not familiar with Boehm GC, but are you really saying that a GC implemented in C does not have a way to drop its roots and ... trigger GC? I find that very surprising.

That said, I read now in [the description](https://github.com/bdwgc/bdwgc/tree/e8e5a4eb19d21c94b14c158ff39cc68a3e105d30#general-description):

> Since the collector does not require pointers to be tagged, it **does not attempt to ensure that all inaccessible storage is reclaimed**.

...

> Unlike manually introduced leaks, the amount of unreclaimed memory typically stays bounded.

Well, looks like it doesn't stay bounded to any sensible amount in practice. I read this as "even in absence of leaks, the amount of memory used is unbounded". That does not seem like a great idea to me.

---

In the face of that, I see why you're suggesting an `exec()`.

But it seems simpler that `nix-build` could start a `nix-instantiate` process that outputs the `.drv`, and then run the `nix-build` code on that. But as Eelco commented in https://github.com/NixOS/nix/issues/5200#issuecomment-910700238, neither that nor the `exec()` will help Import-From-Derivation (IFD).

I don't see how IFD can be implemented with reasonable memory use at all when using an append-only memory-growth GC.

Is there no GC that can actually collect garbage?

**Comment by andir** [comment]
> Specifically, what I'm thinking of is having some special/reserved `argv[0]` values that would get forwarded to the required functions. By re-execing we start with a clean slate and the whole heap would get garbage collected in an efficient manner. Keeping the store paths for derivations in environment variables should be sufficient for the GC (the nix store GC) to keep those alive as roots.
> 
> [@andir](https://github.com/andir) does that make sense to you? This would also significantly improve build performance in the current architecture when not going through the daemon, where the builder processes are `fork`-ed directly from the nix cli process or the daemon process (which might have a gigantic heap and lead to incredibly slow fork() - on the order of 100s of milliseconds). This of course could be further improved by doing more `vfork`-style things or a zygote process, but both of those are complicated and are only tangentially related to this issue.

Yeah, that is something that had been floated  years ago when I first opened my PR. I'm not happy with the fact that we can't really control Boehm GC (my hack of GC'ing several times in a loop worked somewhat for  after cleaning up a bunch of references on the stack). Forking is probably the best solution to the problem even if it is one that I really dislike. It doesn't punish writing bade memory management code.. In this case we can either ditch Boehm GC for something else (what even ist there?) or fork..

**Comment by nh2** [comment]
Having a quick fork-before-instantiate, or `nix-store -r $(nix-instantiate ...)` interim solution seems good to me; since nixpkgs doesn't do IFD, it would already be a big improvement to many NixOS users' RAM problems on `nixos-rebuild`.

**Comment by xokdvium** [comment]
> Forking is probably the best solution to the problem even if it is one that I really dislike.

Now forking without exec might not be feasible anymore btw, since we've started using parallel marking and Boehm is probably not going to be happy about fork() (because the marking threads would be gone). And a that point re-exec might be the less painful solution.
